File: frameworks/webonyx-graphql-php/tests/ResolverTest.php
<?php

declare(strict_types=1);

namespace VelocityBench\Tests;

use PHPUnit\Framework\TestCase;

require_once __DIR__ . '/TestFactory.php';

class ResolverTest extends TestCase
{
    private TestFactory $factory;

    protected function setUp(): void
    {
        $this->factory = new TestFactory();
    }

    protected function tearDown(): void
    {
        $this->factory->reset();
    }

    // ============================================================================
    // User Query Tests
    // ============================================================================

    public function testQueryUserByUuid(): void
    {
        $user = $this->factory->createUser('alice', 'alice@example.com', 'Alice Smith', 'Hello!');

        $result = $this->factory->getUser($user->id);

        $this->assertNotNull($result);
        $this->assertEquals($user->id, $result->id);
        $this->assertEquals('alice', $result->username);
        $this->assertEquals('Alice Smith', $result->fullName);
        $this->assertEquals('Hello!', $result->bio);
    }

    public function testQueryUsersReturnsList(): void
    {
        $this->factory->createUser('alice', 'alice@example.com', 'Alice');
        $this->factory->createUser('bob', 'bob@example.com', 'Bob');
        $this->factory->createUser('charlie', 'charlie@example.com', 'Charlie');

        $users = $this->factory->getAllUsers();

        $this->assertCount(3, $users);
    }

    public function testQueryUserNotFound(): void
    {
        $result = $this->factory->getUser('non-existent-id');

        $this->assertNull($result);
    }

    // ============================================================================
    // Post Query Tests
    // ============================================================================

    public function testQueryPostById(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($user->id, 'Test Post', 'Test content');

        $result = $this->factory->getPost($post->id);

        $this->assertNotNull($result);
        $this->assertEquals('Test Post', $result->title);
        $this->assertEquals('Test content', $result->content);
    }

    public function testQueryPostsByAuthor(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');
        $this->factory->createPost($user->id, 'Post 1', 'Content 1');
        $this->factory->createPost($user->id, 'Post 2', 'Content 2');

        $posts = $this->factory->getPostsByAuthor($user->pkUser);

        $this->assertCount(2, $posts);
    }

    // ============================================================================
    // Comment Query Tests
    // ============================================================================

    public function testQueryCommentById(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Test Post', 'Content');
        $commenter = $this->factory->createUser('commenter', 'commenter@example.com', 'Commenter');
        $comment = $this->factory->createComment($commenter->id, $post->id, 'Great post!');

        $result = $this->factory->getComment($comment->id);

        $this->assertNotNull($result);
        $this->assertEquals('Great post!', $result->content);
    }

    public function testQueryCommentsByPost(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Test Post', 'Content');
        $commenter = $this->factory->createUser('commenter', 'commenter@example.com', 'Commenter');
        $this->factory->createComment($commenter->id, $post->id, 'Comment 1');
        $this->factory->createComment($commenter->id, $post->id, 'Comment 2');

        $comments = $this->factory->getCommentsByPost($post->pkPost);

        $this->assertCount(2, $comments);
    }

    // ============================================================================
    // Relationship Tests
    // ============================================================================

    public function testUserPostsRelationship(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post1 = $this->factory->createPost($user->id, 'Post 1', 'Content 1');
        $post2 = $this->factory->createPost($user->id, 'Post 2', 'Content 2');

        $posts = $this->factory->getPostsByAuthor($user->pkUser);

        $this->assertCount(2, $posts);
        $postIds = array_map(fn($p) => $p->id, $posts);
        $this->assertContains($post1->id, $postIds);
        $this->assertContains($post2->id, $postIds);
    }

    public function testPostAuthorRelationship(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Test Post', 'Content');

        $this->assertNotNull($post->author);
        $this->assertEquals($author->pkUser, $post->author->pkUser);
    }

    public function testCommentAuthorRelationship(): void
    {
        $author = $this->factory->createUser('author', 'author@example.com', 'Author');
        $post = $this->factory->createPost($author->id, 'Test Post', 'Content');
        $commenter = $this->factory->createUser('commenter', 'commenter@example.com', 'Commenter');
        $comment = $this->factory->createComment($commenter->id, $post->id, 'Great!');

        $this->assertNotNull($comment->author);
        $this->assertEquals($commenter->pkUser, $comment->author->pkUser);
    }

    // ============================================================================
    // Edge Case Tests
    // ============================================================================

    public function testNullBio(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');

        $this->assertNull($user->bio);
    }

    public function testEmptyPostsList(): void
    {
        $user = $this->factory->createUser('newuser', 'new@example.com', 'New User');

        $posts = $this->factory->getPostsByAuthor($user->pkUser);

        $this->assertEmpty($posts);
    }

    public function testSpecialCharactersInContent(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');
        $specialContent = "Test with 'quotes' and \"double quotes\" and <html>";
        $post = $this->factory->createPost($user->id, 'Special', $specialContent);

        $this->assertEquals($specialContent, $post->content);
    }

    public function testUnicodeContent(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');
        $unicodeContent = 'Test with émojis 🎉 and ñ and 中文';
        $post = $this->factory->createPost($user->id, 'Unicode', $unicodeContent);

        $this->assertEquals($unicodeContent, $post->content);
    }

    // ============================================================================
    // Performance Tests
    // ============================================================================

    public function testCreateManyPosts(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');

        for ($i = 0; $i < 50; $i++) {
            $this->factory->createPost($user->id, "Post {$i}", 'Content');
        }

        $posts = $this->factory->getPostsByAuthor($user->pkUser);
        $this->assertCount(50, $posts);
    }

    public function testReset(): void
    {
        $this->factory->createUser('user1', 'user1@example.com', 'User 1');
        $this->factory->createUser('user2', 'user2@example.com', 'User 2');

        $this->factory->reset();

        $this->assertEmpty($this->factory->getAllUsers());
    }

    // ============================================================================
    // Validation Tests
    // ============================================================================

    public function testValidUuid(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');

        $this->assertTrue(ValidationHelper::isValidUuid($user->id));
    }

    public function testCreatePostWithInvalidAuthor(): void
    {
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('Author not found');

        $this->factory->createPost('invalid-author', 'Test', 'Content');
    }

    public function testCreateCommentWithInvalidPost(): void
    {
        $user = $this->factory->createUser('user', 'user@example.com', 'User');

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('Post not found');

        $this->factory->createComment($user->id, 'invalid-post', 'Content');
    }

    public function testLongContent(): void
    {
        $user = $this->factory->createUser('author', 'author@example.com', 'Author');
        $longContent = DataGenerator::generateLongString(100000);
        $post = $this->factory->createPost($user->id, 'Long', $longContent);

        $this->assertEquals(100000, strlen($post->content));
    }
}

File: frameworks/rails/test/resolver_test.rb
# frozen_string_literal: true

require_relative 'test_helper'

class ResolverTest < Minitest::Test
  def setup
    @factory = TestFactory.new
  end

  def teardown
    @factory.reset
  end

  # ============================================================================
  # User Query Tests
  # ============================================================================

  def test_query_user_by_uuid
    user = @factory.create_user('alice', 'alice@example.com', 'Alice Smith', 'Hello!')

    result = @factory.get_user(user[:id])

    assert result
    assert_equal user[:id], result[:id]
    assert_equal 'alice', result[:username]
    assert_equal 'Alice Smith', result[:full_name]
    assert_equal 'Hello!', result[:bio]
  end

  def test_query_users_returns_list
    @factory.create_user('alice', 'alice@example.com', 'Alice')
    @factory.create_user('bob', 'bob@example.com', 'Bob')
    @factory.create_user('charlie', 'charlie@example.com', 'Charlie')

    users = @factory.get_all_users

    assert_equal 3, users.length
  end

  def test_query_user_not_found
    result = @factory.get_user('non-existent-id')

    assert_nil result
  end

  # ============================================================================
  # Post Query Tests
  # ============================================================================

  def test_query_post_by_id
    user = @factory.create_user('author', 'author@example.com', 'Author')
    post = @factory.create_post(user[:id], 'Test Post', 'Test content')

    result = @factory.get_post(post[:id])

    assert result
    assert_equal 'Test Post', result[:title]
    assert_equal 'Test content', result[:content]
  end

  def test_query_posts_by_author
    user = @factory.create_user('author', 'author@example.com', 'Author')
    @factory.create_post(user[:id], 'Post 1', 'Content 1')
    @factory.create_post(user[:id], 'Post 2', 'Content 2')

    posts = @factory.get_posts_by_author(user[:pk_user])

    assert_equal 2, posts.length
  end

  # ============================================================================
  # Comment Query Tests
  # ============================================================================

  def test_query_comment_by_id
    author = @factory.create_user('author', 'author@example.com', 'Author')
    post = @factory.create_post(author[:id], 'Test Post', 'Content')
    commenter = @factory.create_user('commenter', 'commenter@example.com', 'Commenter')
    comment = @factory.create_comment(commenter[:id], post[:id], 'Great post!')

    result = @factory.get_comment(comment[:id])

    assert result
    assert_equal 'Great post!', result[:content]
  end

  def test_query_comments_by_post
    author = @factory.create_user('author', 'author@example.com', 'Author')
    post = @factory.create_post(author[:id], 'Test Post', 'Content')
    commenter = @factory.create_user('commenter', 'commenter@example.com', 'Commenter')
    @factory.create_comment(commenter[:id], post[:id], 'Comment 1')
    @factory.create_comment(commenter[:id], post[:id], 'Comment 2')

    comments = @factory.get_comments_by_post(post[:pk_post])

    assert_equal 2, comments.length
  end

  # ============================================================================
  # Relationship Tests
  # ============================================================================

  def test_user_posts_relationship
    user = @factory.create_user('author', 'author@example.com', 'Author')
    post1 = @factory.create_post(user[:id], 'Post 1', 'Content 1')
    post2 = @factory.create_post(user[:id], 'Post 2', 'Content 2')

    posts = @factory.get_posts_by_author(user[:pk_user])

    assert_equal 2, posts.length
    post_ids = posts.map { |p| p[:id] }
    assert_includes post_ids, post1[:id]
    assert_includes post_ids, post2[:id]
  end

  def test_post_author_relationship
    author = @factory.create_user('author', 'author@example.com', 'Author')
    post = @factory.create_post(author[:id], 'Test Post', 'Content')

    assert post[:author]
    assert_equal author[:pk_user], post[:author][:pk_user]
  end

  def test_comment_author_relationship
    author = @factory.create_user('author', 'author@example.com', 'Author')
    post = @factory.create_post(author[:id], 'Test Post', 'Content')
    commenter = @factory.create_user('commenter', 'commenter@example.com', 'Commenter')
    comment = @factory.create_comment(commenter[:id], post[:id], 'Great!')

    assert comment[:author]
    assert_equal commenter[:pk_user], comment[:author][:pk_user]
  end

  # ============================================================================
  # Edge Case Tests
  # ============================================================================

  def test_null_bio
    user = @factory.create_user('user', 'user@example.com', 'User')

    assert_nil user[:bio]
  end

  def test_empty_posts_list
    user = @factory.create_user('newuser', 'new@example.com', 'New User')

    posts = @factory.get_posts_by_author(user[:pk_user])

    assert_empty posts
  end

  def test_special_characters_in_content
    user = @factory.create_user('author', 'author@example.com', 'Author')
    special_content = "Test with 'quotes' and \"double quotes\" and <html>"
    post = @factory.create_post(user[:id], 'Special', special_content)

    assert_equal special_content, post[:content]
  end

  def test_unicode_content
    user = @factory.create_user('author', 'author@example.com', 'Author')
    unicode_content = 'Test with émojis 🎉 and ñ and 中文'
    post = @factory.create_post(user[:id], 'Unicode', unicode_content)

    assert_equal unicode_content, post[:content]
  end

  # ============================================================================
  # Performance Tests
  # ============================================================================

  def test_create_many_posts
    user = @factory.create_user('author', 'author@example.com', 'Author')

    50.times do |i|
      @factory.create_post(user[:id], "Post #{i}", 'Content')
    end

    posts = @factory.get_posts_by_author(user[:pk_user])
    assert_equal 50, posts.length
  end

  def test_reset
    @factory.create_user('user1', 'user1@example.com', 'User 1')
    @factory.create_user('user2', 'user2@example.com', 'User 2')

    @factory.reset

    assert_empty @factory.get_all_users
  end

  # ============================================================================
  # Validation Tests
  # ============================================================================

  def test_valid_uuid
    user = @factory.create_user('user', 'user@example.com', 'User')

    assert ValidationHelper.valid_uuid?(user[:id])
  end

  def test_create_post_with_invalid_author
    assert_raises(RuntimeError) do
      @factory.create_post('invalid-author', 'Test', 'Content')
    end
  end

  def test_create_comment_with_invalid_post
    user = @factory.create_user('user', 'user@example.com', 'User')

    assert_raises(RuntimeError) do
      @factory.create_comment(user[:id], 'invalid-post', 'Content')
    end
  end

  def test_long_content
    user = @factory.create_user('author', 'author@example.com', 'Author')
    long_content = DataGenerator.generate_long_string(100_000)
    post = @factory.create_post(user[:id], 'Long', long_content)

    assert_equal 100_000, post[:content].length
  end
end
